docs/Reminder.md

```diff
@@ -0,0 +1,101 @@
+# Reminder System
+A system that sends out reminders to users based on the standup they have subscribed to.
+
+## Description
+As users create the standups, they can also schedule reminders to remind their subscribers to post their standup notes. In a single standup, only up to 3 reminders can be scheduled.
+The subscribers will receive these reminders according to their preferred mode, e.g., slack or email.
+
+## How it works
+> Example of a stored reminders data in a standup
+```JSON
+{
+    "...": ...,
+
+    "reminders": {
+        "days": [1, 2, 3, 5],
+        "schedules": [
+            {
+                "_id": "625927e2502b0b3ebf9888ba",
+                "time": { "hour": 4, "min": 45 },
+                "notification_time": "2022-04-15T20:45:00.000Z"
+            },
+            {
+                "_id": "625927e2502b0b3ebf9888bb",
+                "time": { "hour": 5, "min": 0 },
+                "notification_time": "2022-04-15T21:00:00.000Z"
+            }
+        ]
+    }
+}
+```
+
+### Sending reminders with a cron job
+1. Reminders are sent through running a cron job
+2. The cron job runs every 3 minutes to query for any standups that have reminders due to be sent out
+    - The query filter to check is - on any notification_time less than the current UTC DateTime
+    - The notification_time field is indexed
+3. If there are any standups with reminders to be sent, they will be dealt with as a batch
+    - It will go through each standup and query all the subscribers (users) associated with that standup to send the reminder.
+    - It will loop through all the reminders in standup to see which reminder notification_time is lesser than the current UTC DateTime.
+    - Afterwhich, it will check if today is a day to send the reminder. In the example above, the `days` field contains `1`, `2`, `3`, and `5`. It means that sending reminders out is only considered on __Monday, Tuesday, Wednesday, and Friday__.
```

Comment:
- I think all days should be considered! so `1` to `7`
- To make the cron job do less work you can first filter the cron jobs by day before step two... So if today is Monday, it would get only standup that have to send reminder on Tuesday! That way it would do less work and be more efficient.
- If a reminder should run everyday what, would it have from `1` to `7`?  I suggest maybe `0` means everyday and if you think that's a good idea please document it here.

Reply:
I was considering this is a company setting.. For each company, it differs in what their working days could be.
If the reminder days cannot be edited by users it means they would receive the reminders even though is their weekends (usually not considered as workdays).

The days written in docs are an example of maybe how a department team in a company would like to receive the reminders on specific days they specify. I was just visualizing in the UI what they could maybe do is through selecting days like [these circles](https://stackoverflow.com/questions/43810161/android-day-picker-like-in-google-clock) and be converted into the array `days` field.

In regards to the second point;
- How the current query works is by checking if any of the reminders is < current UTC DateTime.
- If let's say the selected standup days to send are [1, 2, 3, 4, 5] which means Mon-Fri, on Friday itself it will run and update the `notification_time` field by adding one day to the current DateTime, which will be the DateTime for next day (Saturday).
- Even if Saturday is not a day to send I was thinking that logically the `notification_time` field should be updated to the next date.
- Otherwise, if the `notification_time` field is not updated on the days it is not supposed to be sent out, the date will fall short of two days (Saturday and Sunday) in this case. Then this particular standup will always be queried for, as it is always < current UTC DateTime. As time goes maybe 90% of the standups with reminders in the database will always be queried during every turn of when the cron job runs.
---
- Another way I could think of is to save the `timezone` which is used to generate the corresponding `notification_time`
- Therefore, on any day it should run it will just create a new DateTime in that `timezone` and be converted back to UTC to store. This way even if the reminders `notification_time` field is skipped to be updated throughout the weekends, starting on the new week again, the DateTime will be generated based on the current date, the `timezone`, and the particular `hour` and `min` specified. In this case, the `notification_time` will be free of being dependent on the previous `notification_time` to be updated.
- However, since we agreed not to store any `timezone` the only case I could think of is to update the `notification_time` field for all standups with reminders every single day. Even if users did not select the day to send the reminder out.

Reply:
+1 for the example, I just wanted to make sure you know reminders can be set to run any day of the week.

For the reminders query, I am lost on why you would want to "update" anything to know whether a notification should be sent only two things should happen.

```js
let today = new Date().getDay()
let standups = getAllStandUpsByWeeday(today);

// you can now work with the standups returned, this would avoid making comparison on standups which are supposed to run another day
```

Again, nothing about timezone should matter.

Reply:
Yep, thank you for cross-checking! Will update the docs again to make it clearer.

Ohh! That works too, I'll make the relevant changes and update you again. Thank you for clarifying this 👍🏻

Reply:
@Fenn-CS I was thinking it through again:

If you meant:

1. Changing this [query](https://github.com/huilunang/standupman/blob/standupman-slack/Jobs/Reminder.js#L87) to filter by the days instead, e.g. `'reminders.schedules.days': { $eq: new Date().getDay() }`:
    - This will query will take a longer time as it is returning all the standups with today's day regardless of whether in the current timeframe the standups have reminders to be sent out or not. Moreover, this will be run every 3 minutes which could make these constant reads at a larger scale to be more expensive.
    
or otherwise, if you meant:

2. The issue is not on the query, but after making the query to return all the standups < the current UTC DateTime then further filter it to return all the standups with today's day

Both 2 & 3 I have some concerns, which I will explain in-depth below here and clarify on the "update" part

## The workflow
For instance, in our database, we have these standups:
```JSON
[
    {
        "...": ...,
        
        "reminders": {
            "days": [1, 2, 3, 4, 5],
            "schedules": [
                {
                    "_id": "6259e8f4bec364633724d887",
                    "time": { "hour": 9, "min": 45 },
                    "notification_time": "2022-04-17T01:45:00.000Z"
                },
                {
                    "_id": "6259e8f4bec364633724d888",
                    "time": { "hour": 10, "min": 45 },
                    "notification_time": "2022-04-17T02:45:00.000Z"
                }
            ]
        }
    },
    {
        "...": ...,
        
        "reminders": {
            "days": [1, 2, 3, 4, 5],
            "schedules": [
                {
                    "_id": "6259ea17ca14076352834f47",
                    "time": { "hour": 11, "min": 45 },
                    "notification_time": "2022-04-16T16:45:00.000Z"
                },
                {
                    "_id": "6259ea17ca14076352834f48",
                    "time": { "hour": 12, "min": 45 },
                    "notification_time": "2022-04-16T17:45:00.000Z"
                }
            ]
        }
    },
]
```
Let's say the current UTC DateTime is __`2022-04-16T16:47:00.000Z`__ and the current day is __`Friday`__.

### 1. Getting the standups with reminders
If this [query](https://github.com/huilunang/standupman/blob/standupman-slack/Jobs/Reminder.js#L87) runs:
```js
const date = DateTime.utc();
const standups = await StandUp.find({
    'reminders.schedules.notification_time': { $lt: date },
  })
    .select('name reminders')
    .lean();
```
it should return the standup at `index 1` of the above example.
Since this is based on the DateTime instead, regardless of the days it will just be returned on the query.

<hr>

### 2. Processing the returned standups
- At [here](https://github.com/huilunang/standupman/blob/standupman-slack/Jobs/Reminder.js#L39) it will loop through all the standups returned from the query result.
- Following the procedures, for each standup, it will loop through each of the reminders and it will check if today's a day to send reminders to the users, and then proceed to send those reminders.
- Then lastly it will proceed to update the `notification_time` field to the next day.
    - > __The point is on [here](https://github.com/huilunang/standupman/blob/standupman-slack/Jobs/Reminder.js#L54). Notice that the `notification_time` field is updated regardless of whether today's a day to send the reminder or not.__

### Reason to not filter standup based on days
1. The `notification_time` field is the way to check if the reminders should be sent out at [here](https://github.com/huilunang/standupman/blob/standupman-slack/Jobs/Reminder.js#L43).
    - For instance, continuing from the above part (2), looking just at the reminder with `_id: 6259ea17ca14076352834f47`. The current UTC DateTime is `2022-04-17T16:47:00.000Z` and the day now is `Saturday`.
    - If we were to follow the scenario filtering the standups by day, it means the standup at `index 1` will not fall into being queried. This means that this `notification_time` will stay at `"2022-04-17T16:45:00.000Z"` (Saturday's DateTime) after yesterday's (Friday) update.
    - Moving forward, the current UTC DateTime is `2022-04-19T00:00:00.000Z` and the day now is `Monday`. The reminder `_id: 6259ea17ca14076352834f47` still has its `notification_time` field at `"2022-04-17T16:45:00.000Z"`, which is the Datetime of Saturday's.
    - Today is a day to send out the reminder, it will follow the procedures stated in part (2), and update the `notification_time` by adding a day to it. Therefore, the current reminder `_id: 6259ea17ca14076352834f47` will have its `notification_time` updated to `"2022-04-18T16:45:00.000Z"`.
    - On the next round the Cron job runs, it will detect this particular reminder again as it is less than the current UTC DateTime `2022-04-19T00:03:00.000Z`. These 2 times it runs are making up for the weekends that we did not make the update.
    - Moreover, for the standup at `index 1`, they have 2 reminders. Therefore, to make it up, within the span of ~6 minutes they will be spammed with 4 reminders just so for the DateTime to catch up to today's UTC date.
    - Another issue is also if the `notification_time` is not updated, they will fall into being queried for every 3 minutes the Cron job runs during `Saturday` and `Sunday`.
    - Therefore, this is why regardless of when the days users select for the reminders to be sent out, all standup that has reminders matching the  ___current DateTime period___ should be updated to the next day.
    - Otherwise, every 3 minutes there will always be additional query results returned but not used, and also have these gaps in the date causing unwanted reminders will be sent out.
    ---
    - The reason for using the `notification_time` field to do the check is the only way I could think of to effectively check if at the ___current DateTime period___ that reminder should be sent out or not.
    - Both the `Date` and `Time` (making up the DateTime) in this case are important to make the check.
    - As for each standup, the reminders timings are potentially different since users can opt for which `timezone` they want their reminder time to follow. This results in the generated `notification_time` being different.

Reply:
This makes sense! Thanks for giving more details about this!